core/src/graph/graph.rs

```diff
@@ -569,15 +609,75 @@ impl Graph {
 	/// Determine if page is full
 	///  aggressive:false -> use a simple heuristic based on the number of connections
 	///  aggressive:true  -> do actual compression to determine resulting actual page size
-	pub fn is_page_full(&self, page: &GraphPage, aggressive: bool) -> bool {
+	pub fn try_add_connection_to_page(
+		&self,
+		page: &mut GraphPage,
+		connection_id: &DsnpUserId,
+		aggressive: bool,
+		dsnp_version_config: &DsnpVersionConfig,
+	) -> DsnpGraphResult<()> {
 		let connection_type = self.get_connection_type();
-		let max_connections_per_page = *PAGE_CAPACITIY_MAP.get(&connection_type).unwrap_or(&50);
+		let max_connections_per_page =
+			*PAGE_CAPACITY_MAP.get(&connection_type).unwrap_or_else(|| {
+				let mut capacities: Vec<&usize> = PAGE_CAPACITY_MAP.values().collect();
+				capacities.sort();
+				capacities.first().unwrap() // default: return smallest capacity value
+			});
+		let encryption_key = match self.get_connection_type().privacy_type() {
+			PrivacyType::Public => None,
+			PrivacyType::Private =>
+				self.user_key_manager.read().unwrap().get_resolved_active_key(self.user_id),
+		};
 
-		if !aggressive {
-			return page.connections().len() >= max_connections_per_page
+		// Regardless of whether we're in aggressive mode, if the page is trivially non-full,
+		// just try and add the connection
+		if page.connections().len() < max_connections_per_page {
+			if let Err(e) = page.add_connection(connection_id) {
+				return Err(e)
+			}
+
+			return Ok(())
+		} else if !aggressive {
+			return Err(DsnpGraphError::PageTriviallyFull)
 		}
 
-		todo!()
+		let max_page_size = self.environment.get_config().max_graph_page_size_bytes as usize;
+		let mut temp_page = page.clone();
+		if let Err(e) = temp_page.add_connection(connection_id) {
+			return Err(e)
+		}
+
+		let page_blob = match connection_type {
+			ConnectionType::Follow(PrivacyType::Public) |
+			ConnectionType::Friendship(PrivacyType::Public) => temp_page.to_public_page_data(),
+			ConnectionType::Follow(PrivacyType::Private) => {
+				let encryption_key =
+					encryption_key.as_ref().ok_or(DsnpGraphError::NoResolvedActiveKeyFound)?;
+				temp_page.clear_prids();
+				temp_page.to_private_page_data(dsnp_version_config, &encryption_key)
+			},
+			ConnectionType::Friendship(PrivacyType::Private) => {
+				let encryption_key =
+					encryption_key.as_ref().ok_or(DsnpGraphError::NoResolvedActiveKeyFound)?;
+				self.apply_prids(&mut temp_page, &vec![*connection_id], &encryption_key)
+					.expect("Error applying prids to page");
+				temp_page.to_private_page_data(dsnp_version_config, &encryption_key)
+			},
+		};
+
+		match page_blob {
+			Ok(blob) =>
+				if blob.content.len() > max_page_size {
+					Err(DsnpGraphError::PageAggressivelyFull)
+				} else {
+					if let Err(e) = page.add_connection(connection_id) {
+						return Err(e)
+					}
+
```

Comment:
```suggestion
					page.add_connection(connection_id)?;

```

Reply:
Ah, right... now that we're returning real errors we can do that 👍🏻